core/types.ml

```diff
@@ -649,282 +539,305 @@ let primary_kind_of_type_arg : type_arg -> PrimaryKind.t =
+     self#is_row (StringSet.empty, IntSet.empty, IntSet.empty))
+end
 
-let type_can_be_unl = type_can_be_unl (IntSet.empty, IntSet.empty)
-let row_can_be_unl = row_can_be_unl (IntSet.empty, IntSet.empty)
+class virtual type_iter = object(self)
+  method visit_var : 'a 'c. ([< 'a meta_max_basis > `Var] as 'c) point -> (int * subkind * freedom) -> unit = fun _ _ -> ()
 
-let make_point_unl : ((var_set * var_set) -> 'a -> unit) -> (var_set * var_set) -> [< 'a meta_max_basis] point -> unit =
-  fun f ((rec_vars, quant_vars) as vars) point ->
+  method visit_point : 'a 'c . (predicate_context -> 'a -> unit) -> predicate_context -> ([< 'a meta_max_basis > `Var] as 'c) point -> unit
+    = fun f ((rec_appl, rec_vars, quant_vars) as vars) point ->
     match Unionfind.find point with
     | `Closed -> ()
-    | `Var (v, (lin, _), `Rigid)       -> if IntSet.mem v quant_vars || Linearity.is_nonlinear lin then () else assert false
-    | `Var (var, (_, rest), `Flexible) -> Unionfind.change point (`Var (var, (lin_unl, rest), `Flexible))
+    | `Var ((id, _, _) as var) -> if not (IntSet.mem id quant_vars) then self#visit_var point var
     | `Body t -> f vars t
-    | `Recursive (var, t) ->
-      check_rec var rec_vars () (fun rec_vars' -> f (rec_vars', quant_vars) t)
+    | `Recursive (var, t) -> check_rec var rec_vars () (fun rec_vars' -> f (rec_appl, rec_vars', quant_vars) t)
 
-let rec make_type_unl : var_set * var_set -> typ -> unit =
-  fun ((rec_vars, quant_vars) as vars) ->
-    function
+  method visit_type ((rec_appl, rec_vars, quant_vars) as vars) : typ -> unit = function
     | `Not_typed -> assert false
-    | `Primitive _ | `Function _ | `Table _ | `End | `Application _ | `Effect _ | `Lens _ -> ()
-    | `RecursiveApplication _ -> ()
-    | `Record r | `Variant r -> make_row_unl vars r
-    | `Alias (_, t) -> make_type_unl vars t
-    | `ForAll (qs, t) -> make_type_unl (rec_vars, add_quantified_vars !qs quant_vars) t
-    | `MetaTypeVar point -> make_point_unl make_type_unl vars point
-    | `Dual s -> make_type_unl vars s
-    | _ -> assert false
-and make_field_unl vars =
-  function
-  | `Absent -> ()
-  | `Present t -> make_type_unl vars t
-  | `Var point -> make_point_unl make_field_unl vars point
-and make_row_unl vars (fields, row_var, _) =
-  make_point_unl make_row_unl vars row_var;
-  FieldEnv.iter (fun _name -> make_field_unl vars) fields
+    | `Primitive _ -> ()
+    | `Function (a, e, r) | `Lolli (a, e, r) -> self#visit_type vars a; self#visit_row vars e; self#visit_type vars r
+    | `Record r | `Effect r | `Variant r -> self#visit_row vars r
+    | `Table _ -> ()
+    | `Lens _ -> ()
+    | `Alias (_, t) -> self#visit_type vars t
+    | `MetaTypeVar point -> self#visit_point self#visit_type vars point
+    | `ForAll (qs, t) -> self#visit_type (rec_appl, rec_vars, add_quantified_vars !qs quant_vars) t
+    | `Application _ | `RecursiveApplication _ -> () (* For now, we assume these are acceptable! *)
```

Comment:
Again -- I'm surprised that you have a visitor where the type arguments aren't visited -- why is this?